pop_tools/budget.py

```diff
@@ -0,0 +1,328 @@
+    # integral? In that case, this would break.
+    vadv = (ds.WT.shift(z_t=-1).fillna(0) - ds.WT).isel(z_t=slice(0, -1))
+    vadv = vadv.rename('vadv')
+    vadv = _convert_units(vadv)
+    vadv.attrs['long_name'] = 'vertical advection'
+    return vadv
+
+
+def _compute_vertical_mixing(ds, grid, mask, kmax=None):
+    """Compute contribution from vertical mixing."""
+    ds = _subset_slice_mask(ds, var_list=['DIA_IMPVF', 'KPP_SRC'], kmax=kmax, mask=mask)
+
+    # Only need to flip sign of DIA_IMPVF.
+    ds['DIA_IMPVF'] = _convert_to_tendency(ds['DIA_IMPVF'], grid, sign=-1, force_area=True)
+    ds['KPP_SRC'] = _convert_to_tendency(ds['KPP_SRC'], grid, kmax=kmax)
+
+    # Compute divergence of diapycnal mixing
+    diadiff = _compute_vertical_divergence(ds.DIA_IMPVF)
+
+    # Sum KPP and diapycnal mixing to create vmix term.
+    vmix = (ds.KPP_SRC + diadiff).rename('vmix')
+    vmix = _convert_units(vmix)
+    vmix.attrs['long_name'] = 'vertical mixing'
+    return vmix
+
+
+def _compute_SMS(ds, grid, mask, kmax=None):
+    """Compute SMS term from biology."""
+    ds = _subset_slice_mask(ds, var_list=['SMS'], kmax=kmax, mask=mask)
+    ds = _convert_to_tendency(ds, grid, kmax=kmax)
+    ds = _convert_units(ds)
+    ds = ds.rename({'SMS': 'sms'})
+    ds.attrs['long_name'] = 'source/sink'
+    return ds
+
+
+def _compute_surface_flux(ds, grid, mask):
+    """Computes virtual fluxes of tracer."""
+    ds = _subset_slice_mask(ds, var_list=['STF'], mask=mask)
+    ds = _convert_to_tendency(ds, grid)
+    ds = _convert_units(ds)
+    stf = ds.rename({'STF': 'stf'})
+    stf.attrs['long_name'] = 'surface tracer flux'
+    return stf
+
+
+def _compute_virtual_flux(ds, grid, mask):
+    """Computes virtual fluxes of tracer."""
+    ds = _subset_slice_mask(ds, var_list=['FvICE', 'FvPER'], mask=mask)
+    ds = _convert_to_tendency(ds, grid)
+    ds = _convert_units(ds)
+    vf = (ds.FvICE + ds.FvPER).rename('vf')
+    vf.attrs['long_name'] = 'virtual flux'
+    return vf
+
+
+def _process_input_dataset(ds):
+    """Checks that input dataset has appropriate variables, etc."""
+    mandatory_vars = ['UE', 'VN', 'WT', 'HDIFE', 'HDIFN', 'HDIFB', 'DIA_IMPVF', 'KPP_SRC']
+    if not set(mandatory_vars).issubset(ds):
+        missing_vars = [v for v in mandatory_vars if v not in ds]
+        error_msg = (
+            'Input dataset does not contain the mandatory variables for budget analysis. '
+            + f'`ds` is missing {missing_vars}'
+        )
+        raise IOError(error_msg)
+
+    # Drop coordinates for shift, roll, etc.
+    coord_vars = [c for c in ds.coords if c != 'time']
+    ds = ds.drop(coord_vars)
+    if 'z_w_top' in ds.dims:
+        ds = ds.drop('z_w_top')
+        ds = ds.rename({'z_w_top': 'z_t'})
+    if 'z_w_bot' in ds.dims:
+        ds = ds.drop('z_w_bot')
+        ds = ds.rename({'z_w_bot': 'z_t'})
+
+    # Force datetime
+    if not np.issubdtype(ds.time.dtype, np.datetime64):
```

Comment:
@bradyrx, we can extend this to support datasets with a time coordinate decoded using `cftime._cftime.datetime`  which extends to calendars other than the proleptic Gregorial calendar.

Xarray now supports `cftime._cftime.datetime` based data types for groupby and resample operations to the same extent as for `datetime`/`numpy.datetime` based data types. 




```python
In [26]: ds.time
Out[26]: 
<xarray.DataArray 'time' (time: 24)>
array([cftime.DatetimeNoLeap(1, 2, 1, 0, 0, 0, 0, 4, 32),
       cftime.DatetimeNoLeap(1, 3, 1, 0, 0, 0, 0, 4, 60),
       cftime.DatetimeNoLeap(1, 4, 1, 0, 0, 0, 0, 0, 91),
       cftime.DatetimeNoLeap(1, 5, 1, 0, 0, 0, 0, 2, 121),
       cftime.DatetimeNoLeap(1, 6, 1, 0, 0, 0, 0, 5, 152),
       cftime.DatetimeNoLeap(1, 7, 1, 0, 0, 0, 0, 0, 182),
      .........
       cftime.DatetimeNoLeap(2, 9, 1, 0, 0, 0, 0, 0, 244),
       cftime.DatetimeNoLeap(2, 10, 1, 0, 0, 0, 0, 2, 274),
       cftime.DatetimeNoLeap(2, 11, 1, 0, 0, 0, 0, 5, 305),
       cftime.DatetimeNoLeap(2, 12, 1, 0, 0, 0, 0, 0, 335),
       cftime.DatetimeNoLeap(3, 1, 1, 0, 0, 0, 0, 3, 1)], dtype=object)
Coordinates:
  * time     (time) object 0001-02-01 00:00:00 ... 0003-01-01 00:00:00
Attributes:
    long_name:  model time
    bounds:     time_bounds
```

Cftime time values are stored in numpy array with dtype='object'. 

```python
In [27]: ds.time.dtype
Out[27]: dtype('O')
```

The only way to test for cftime class that I know of, consists of retrieving one value from the array :

```python
In [28]: isinstance(ds.time.data[0], cftime._cftime.datetime)
Out[28]: True
```

Reply:
Note from a conversation with @mclong, this will be better moved to a helper function to help users construct the dataset. I think this was only here to handle temporal resampling, which will be taken away in this PR.